### PR TITLE
Feature/issue 3 create crud blog

### PR DESCRIPTION
**Pull request by mdrakibtrofder**
Open a pull request for Feature/issue 3 create crud blog branch

### CODE REVIEW

File: services/blog.service.js [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+const getAllBlogs = () => {
+  return blogRepository.getAllBlogs();
+};
+
+const createBlog = async (body) => {
+  const id = commonUtil.generateUUID();
+  const title = body.title;
+  const description = body.description;
+  const userName = String(body.username);
```
Comment (rafaelragib) [inline services/blog.service.js]: fix variable name. username isn't two words for that don't use camelcase

File: controllers/blog.controller.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+    res.send(response);
+
+  } catch (err) {
+    console.error(err);
+    res.send({ status: 500, message: "Internal Server Error" });
+  }
+};
+
+const getBlogByBlogId = async (req, res) => {
```
Comment (rafaelragib) [inline controllers/blog.controller.js]: no need to mention blogid in the name. just write simple gitBlogById.

File: repositories/blog.repository.js [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+  try {
+    const blogs = await Blog.findAll({
+      where: {
+        authorId: authorId,
+      },
+    });
+
+    if (blogs) {
+      return blogs;
+    }
```
Comment (rafaelragib) [inline repositories/blog.repository.js]: what will happen if the authorId doesn't match any data? wouldn't it response an empty array?

File: services/blog.service.js [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+  if(!description&&!title){
+    return { status: 400, message: `Both description and title fields are empty` };
+
+  }
+    if(!description){
+    description = result.description;
+  }
+  else if(!title){
+    title = result.title;
+  }
```
Comment (rafaelragib) [inline services/blog.service.js:95]: how do you resassign const variable?

File: controllers/auth.controller.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
+const authService = require("../services/auth.service");
+require("dotenv").config();
+
+const registerUser = async (req, res) => {
+
+  try {
+    if (JSON.stringify(req.body)==="{}") {
```
Comment (rafaelragib) [inline controllers/auth.controller.js]: JSON.stringify is a costly function. you can use Object.keys which is much faster.

**Review by rafaelragib — CHANGES_REQUESTED** [review]

File: repositories/blog.repository.js (hunk 3 above)
Comment (mdrakibtrofder) [reply repositories/blog.repository.js]: I send null value in blog service and check if response is null then send data is not found

File: services/blog.service.js (hunk 4 above)
Comment (mdrakibtrofder) [reply services/blog.service.js:95]: I have changed it to let.

File: controllers/auth.controller.js (hunk 5 above)
Comment (mdrakibtrofder) [reply controllers/auth.controller.js]: Yes, I have used Object.keys.

File: controllers/auth.controller.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+const registerUser = async (req, res) => {
+  if (!Object.keys(req.body).length) {
+    return res.send({ status: 400, message: "Request body is empty" });
+  }
+
+  try {
+     const response = await authService.registerUser(req.body);
+
+    if (response.status==201) {
```
Comment (rafaelragib) [inline controllers/auth.controller.js:12]:
```suggestion
    if (response.status===201) {
```

File: controllers/auth.controller.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+const loginUser = async (req, res) => {
+  if (!Object.keys(req.body).length) {
+    return res.send({ status: 400, message: "Request body is empty" });
+  }
+
+  try {
+   
+    const response = await authService.loginUser(req.body);
+
+   if (response.status==201) {
```
Comment (rafaelragib) [inline controllers/auth.controller.js:37]: loginuser should return 200 status

File: controllers/blog.controller.js [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+  if (!Object.keys(req.body).length) {
+    return res.send({ status: 400, message: "Request body is empty" });
+  }
+  try {
+    if (!req.params.id) {
+      res.send({ status: 400, message: "Request parameter is empty" });
+    }
+    const updateBlogByIdResponse =
+      await blogService.updateBlogById(req.body, req.params.id);
+    res.send({status: updateBlogByIdResponse.status, message:updateBlogById.message});
```
Comment (rafaelragib) [inline controllers/blog.controller.js:77]:
```suggestion
    res.send({status: updateBlogByIdResponse.status, message:updateBlogByIdResponse.message});
```
shouldn't this be updateBlogByIdResponse?

File: services/blog.service.js [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+};
+
+const getBlogByAuthorId = async (authorId) => {
+  const result = await blogRepository.getBlogByAuthorId(authorId);
+
+  if (!result) {
+    return {
+      status: 404,
+      message: `This blog is not found in database`,
```
Comment (rafaelragib) [inline services/blog.service.js:65]: No blog found with this author id

File: services/blog.service.js [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+const updateBlogById = async (body, blogId) => {
+  let title = body.title;
+  let description = body.description;
+
+
+  const result = await blogRepository.getBlogById(blogId);
+  if (!result) {
+    return {
+      status: 404,
+      message: `This blog is not found in database`,
```
Comment (rafaelragib) [inline services/blog.service.js:81]: No blog found with this id

File: services/blog.service.js [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+  const result = await blogRepository.getBlogById(blogId);
+
+  if (!result) {
+    return {
+      status: 404,
+      message: `${blogId} is not found in database`,
+    };
+  }
+
+  const isUserDeleted = blogRepository.deleteBlogById(blogId);
```
Comment (rafaelragib) [inline services/blog.service.js:123]: await keyword missing

File: utils/blog.util.js [hunk 12]
```diff
@@ -0,0 +1,28 @@
+const jwt = require("jsonwebtoken");
+const userService = require("../services/user.service");
+
+
+const getAuthorId = async(req) =>{
+  try {
+    const accessToken = cookies.jsontoken;
```
Comment (rafaelragib) [inline utils/blog.util.js:7]: from where are you getting the cookies.jsontoken?

File: utils/blog.util.js [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+
+    if (!accessToken) {
+      return res.status(403).json({
+        message: "JWT is Expired",
+      });
```
Comment (rafaelragib) [inline utils/blog.util.js:12]: from where are you getting the `res`?

File: utils/validation.util.js [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
+    return emailRegex.test(email);
+  };
+  
+const checkPasswordLength = (password) => {
+    if (password.length >= 6) {
+      return true;
+    } else {
+      return false;
+    }
```
Comment (rafaelragib) [inline utils/validation.util.js:18]: `return password.length >= 6`

it can be written in only one statement

File: utils/validation.util.js [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+  };
+  const generateHashPassword = async (plainPassword) => {
+    const salt = await bcrypt.genSalt(10);
+    const password = await bcrypt.hash(plainPassword, salt);
```
Comment (rafaelragib) [inline utils/validation.util.js:22]:
```suggestion
    const hashPassword = await bcrypt.hash(plainPassword, salt);
```

File: utils/validation.util.js [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+    return password;
+  };
```
Comment (rafaelragib) [inline utils/validation.util.js:24]: this should have a try catch block

**Review by rafaelragib — CHANGES_REQUESTED** [review]

**Pull request merged**